src/services/datamigration/datamigrationService.js

```diff
@@ -0,0 +1,225 @@
+							reject(err);
+						}
+						
+						const topickeypairs = [];
+
+						keys.forEach((key) => {
+							topickeypairs.push({ topic: topic, key: key });
+						});
+
+						resolve(topickeypairs);
+					});
+				}));
+
+				Promise.all(keysPromises).then((allpairs) => {
+					const pairs = allpairs.reduce((acc, curr) => {
+						return [...acc, ...curr];
+					});
+		
+					const valuesPromises = [];
+		
+					pairs.forEach((pair) => {
+						valuesPromises.push(new Promise((resolve, reject) => {
+							StorageClient.get({ topic: pair.topic, key: pair.key }, (err, value) => {
+								if (err) {
+									reject(err);
+								}
+								resolve({topic: pair.topic, key: pair.key, value: value});
+							});
+						}));
+					});
+		
+					Promise.all(valuesPromises).then((data) => {
+						this.saveAllData(topics, data);
+					});
+				});
+			});
+		})
+		
+		
+	}
+
+	/**
+	 * saveAllData
+	 * 
+	 * Saves data from the old storage adapter into the new one.
+	 * 
+	 * Invoked when all data has been retrieved from the old adapter.
+	 * @param {Array} topics Array of all topics retrieved.
+	 * @param {Array} data Array of all data received from retrieveDatabase
+	 */
+	saveAllData(topics, dataset) {
+		const promiseSet = [];
+
+		topics.forEach((topic) => {
+			promiseSet.push(new Promise((resolve, reject) => {
+				// Reset to use new storage adapter
+				StorageClient.setStore({ topic: topic, dataStore: this.newAdapter }, (err, data) => {
+					if (err) {
+						reject(new Error(err));
+					}
+					resolve(data);
+				});
+			}));
+		});
+
+		Promise.all(promiseSet).then(() => {
+			const saveSet = [];
+			
+			saveSet.push(new Promise((resolve, reject) => {
+
+				dataset.forEach((datum) => {
+					StorageClient.save({ topic: datum.topic, key: datum.key, value: datum.value }, (err, response) => {
+						if (err) {
+							reject(new Error(err));
+						}
+						resolve(response);
+					});
+				});
+			}));
+
+			Promise.all(saveSet).then(() => {
+				StorageClient.save({ topic: "finsemble", key: `migrated_from_${this.adapter}`, value: Date.now() });
+				RouterClient.publish("Migration", "end");
+			});
+		});
+	}
+
+	/**
+	 * fetchUserStatus
+	 * 
+	 * Determines if the user has been migrated from IndexedDBAdapter (or another storage adapter)
+	 * by checking for a key in the new storage service of "migrated_from_<adapter to migrate from>", 
+	 * which in this example is "migrated_from_IndexedDBAdapter".
+	 */
+	fetchUserStatus() {
+		StorageClient.get({ topic: "finsemble", key: `migrated_from_${this.adapter}` }, (err, data) => {
+			if (err) {
+				throw new Error(err);
+			}
+
+			if (!data) {
+				// The user does not have a migration record.
+				const promiseSet = [];
+
+				StorageClient.setStore({ topic: "finsemble", dataStore: this.adapter }, (err, data) => {
+					if (err) {
+						reject(new Error(err));
+					}
+					StorageClient.keys({ topic: "finsemble" }, (err, keys) => {
+						if (err) {
+							throw new Error(err);
```

Comment:
Throwing an error in a service probably isn't a good idea. You should log the error and return.